File: WildcardMatching.java
public class Solution {
	private boolean matchFrom(int i, int len, String s, String p) {
		int n = p.length();
		if (i + n > len)
			return false;
		for (int j = 0; j < n; j++)
			if (p.charAt(j) != '?' && p.charAt(j) != s.charAt(i + j))
				return false;
		return true;
	}

	public boolean isMatch(String s, String p) {
		int len = s.length();
		String[] chunks = p.split("\\*+", -1);
		int cnt = chunks.length;
		if (cnt == 1)// no stars
			return s.length() == p.length() && matchFrom(0, len, s, p);
		int l = chunks[0].length();
		int r = chunks[cnt - 1].length();
		if (l + r > len || !matchFrom(0, len, s, chunks[0])
				|| !matchFrom(len - r, len, s, chunks[cnt - 1]))
			return false;
		int begin = l;
		len -= r;
		for (int i = 1; i < cnt - 1; i++) {
			while (begin < len && !matchFrom(begin, len, s, chunks[i]))
				begin++;
			if (begin == len)
				return false;
			begin += chunks[i].length();
		}
		return true;
	}
}

File: WildcardMatching.py
class Solution:
    def isMatch(self, s, p):
        def match_from(i,n, s, p):
            m = len(p)
            if i+m > n:return False
            return reduce(lambda x,y: x and y,[p[j]=='?' or p[j]==s[i+j] for j in xrange(m)], True)
        chunks = re.split('\*+',p)
        cnt,n = len(chunks),len(s)
        if cnt == 1:
            return len(s)==len(p) and match_from(0,n,s,p)
        l,r = len(chunks[0]),len(chunks[-1])
        if l + r > n or not match_from(0,l,s,chunks[0]) or not match_from(n-r,n,s,chunks[-1]):
            return False
        begin = l
        n -= r
        for i in xrange(1,cnt - 1):
            chunk = chunks[i]
            while begin<n and (not match_from(begin, n, s,chunk)):
                begin += 1
            if begin == n:
                return False
            begin += len(chunk)
        return True
